cli/azd/internal/repository/initializer.go

```diff
@@ -0,0 +1,240 @@
+	Initialize(ctx context.Context, templateUrl string, templateBranch string) error
+	InitializeEmpty(ctx context.Context) error
+}
+
+type initializer struct {
+	azdCtx  *azdcontext.AzdContext
+	console input.Console
+	gitCli  git.GitCli
+}
+
+func NewInitializer(azdCtx *azdcontext.AzdContext,
+	console input.Console,
+	gitCli git.GitCli) Initializer {
+	return &initializer{
+		azdCtx:  azdCtx,
+		console: console,
+		gitCli:  gitCli,
+	}
+}
+
+func (i *initializer) Initialize(ctx context.Context, templateUrl string, templateBranch string) error {
+	var err error
+	stepMessage := fmt.Sprintf("Downloading template code to: %s", output.WithLinkFormat("%s", i.azdCtx.ProjectDirectory()))
+	i.console.ShowSpinner(ctx, stepMessage, input.Step)
+	defer i.console.StopSpinner(ctx, stepMessage+"\n", input.GetStepResultFormat(err))
+
+	staging, err := os.MkdirTemp("", "az-dev-template")
+
+	if err != nil {
+		return fmt.Errorf("creating temp folder: %w", err)
+	}
+
+	// Attempt to remove the temporary directory we cloned the template into, but don't fail the
+	// overall operation if we can't.
+	defer func() {
+		_ = os.RemoveAll(staging)
+	}()
+
+	target := i.azdCtx.ProjectDirectory()
+
+	err = i.gitCli.FetchCode(ctx, templateUrl, templateBranch, staging)
+	if err != nil {
+		return fmt.Errorf("\nfetching template: %w", err)
+	}
+
+	log.Printf(
+		"template init, checking for duplicates. source: %s target: %s",
+		staging,
+		target,
+	)
+
+	duplicateFiles, err := determineDuplicates(staging, target)
+	if err != nil {
+		return fmt.Errorf("checking for overwrites: %w", err)
+	}
+
+	if len(duplicateFiles) > 0 {
+		i.console.Message(ctx, "warning: the following files will be overwritten with the versions from the template:")
+		for _, file := range duplicateFiles {
+			i.console.Message(ctx, fmt.Sprintf(" * %s\n", file))
+		}
+
+		overwrite, err := i.console.Confirm(ctx, input.ConsoleOptions{
+			Message:      "Overwrite files with versions from template?",
+			DefaultValue: false,
+		})
+
+		if err != nil {
+			return fmt.Errorf("prompting to overwrite: %w", err)
+		}
+
+		if !overwrite {
+			return errors.New("confirmation declined")
+		}
+	}
+
+	if err := copy.Copy(staging, target); err != nil {
+		return fmt.Errorf("copying template contents: %w", err)
+	}
+
+	i.console.StopSpinner(ctx, stepMessage+"\n", input.GetStepResultFormat(err))
+	err = i.writeAzdAssets(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Initializes an empty (bare minimum) azd repository.
+func (i *initializer) InitializeEmpty(ctx context.Context) error {
+	return i.writeAzdAssets(ctx)
+}
+
+func (i *initializer) writeAzdAssets(ctx context.Context) error {
+	// Check to see if `azure.yaml` exists, and if it doesn't, create it.
+	if _, err := os.Stat(i.azdCtx.ProjectPath()); errors.Is(err, os.ErrNotExist) {
+		stepMessage := fmt.Sprintf("Creating a new %s file.", azdcontext.ProjectFileName)
+
+		i.console.ShowSpinner(ctx, stepMessage, input.Step)
```

Comment:
Suspect this is just code motion of existing code but it's a little weird that we use the step logic here (especially since the action, as far as IO bound actions go will complete quite fast) and yet we don't do it for other things like creating the `.azure` folder or the `.gitignore` work.

Just something I noticed.

Reply:
Agreed. Added to #1106 to discuss how we should deal with it.